scrapy_poet/middleware.py

```diff
@@ -48,11 +51,16 @@ def __init__(self, crawler: Crawler) -> None:
     def from_crawler(cls: Type[InjectionMiddlewareTV], crawler: Crawler) -> InjectionMiddlewareTV:
         o = cls(crawler)
         crawler.signals.connect(o.spider_closed, signal=signals.spider_closed)
+        crawler.signals.connect(o.spider_opened, signal=signals.spider_opened)
         return o
 
     def spider_closed(self, spider: Spider) -> None:
         self.injector.close()
 
+    def spider_opened(self, spider):
+        scrapy_downloader_var.set(spider.crawler.engine.download)
```

Comment:
How does it work with multiple spiders in the same process? It seems this might be a problem. I'm not sure about the solution yet.

Reply:
I've tested this out but couldn't find any problem as running multiple spiders in the same process works okay.

@kmike I've tagged you on a separate PR which shows the simple test I've made. Perhaps there's some problem or behavior change that I've missed out.

Reply:
answered there, thanks @BurnzZ!

Reply:
Hi @kmike , cheers again for raising. 

After some deep investigation on this matter, it would seem that **Twisted** only has a limited support for **ContextVars** at the moment. It currently breaks some of our expectations in using **ContextVars** inside Scrapy. _(see section below for full explanation)_

With that, I've updated the approach inside `scrapy-poet` to use **Dependency Injection** instead. This creates a consistent non-leaking value across different multiple CrawlerProcesses. Here's the commit for this: https://github.com/scrapinghub/scrapy-poet/pull/62/commits/e3b96dce17b199ebe3804aa68f20a1e6ee3b96d0

### Investigation Notes
---------------------

It would seem that having this order of multiple spiders in a single process:

```python
process = CrawlerProcess(settings)
process.crawl(SpiderAAA)
process.crawl(SpiderBBB)
process.crawl(SpiderCCC)
process.start()
```

would result in:

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ❌ **SpiderBBB** using the downloader from **SpiderAAA**
- ❌ **SpiderCCC** using the downloader from **SpiderAAA**

In other words, when the first spider sets the downloader, it would be used throughout in the succeeding spiders, regardless if the succeeding spiders attempt to update the ContextVars directly. I've tried it with more spiders and the behavior is simply the same.

We can override the `ContextVar` value inside the spider though. However, after some testing, it would seem that Twisted leaks the context outside of each crawler process. So if we update the `ContextVar` value inside **SpiderBBB**, it would leak as:

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ✅ **SpiderBBB** using the downloader from **SpiderBBB**
- ❌ **SpiderCCC** using the downloader from **SpiderBBB**


Interestingly, the behavior above only happens when `twisted>=21.2.0` since it introduced this change:
> twisted.internet.defer.inlineCallbacks and ensureDeferred will now associate a contextvars.Context with the coroutines they run, meaning that ContextVar objects will maintain their value within the same coroutine, similarly to asyncio Tasks. This functionality requires Python 3.7+, or the contextvars PyPI backport to be installed for Python 3.5-3.6. (#9719, #9826)

Here's the code comment in Twisted that tests the ContextVars in InlineCallbacks ([reference](https://github.com/twisted/twisted/pull/1192/files#diff-d51be359d7730f86734b450ecbf6dc8f54090e84e7efcf5616399fb5cffc96a7R3114-R3118)).

> When an inlineCallbacks function is called, the context is taken from
> when it was first called. When it resumes, the same context is applied.

A twisted version below `21.2.0` would result in the the following as expected ([Reference](https://github.com/twisted/twisted/releases/tag/twisted-21.2.0)):

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ✅ **SpiderBBB** using the downloader from **SpiderBBB**
- ✅ **SpiderCCC** using the downloader from **SpiderCCC**

However, the expected correct behavior above only happens during normal spider operation. When the time comes where an Additional Request is to be executed in the downloader, the ContextVar raises a `LookupError` since the context is not preserved. ❌

After digging further, it would seem that Twisted doesn't support all of the ContextVar features completely:

- [Support contextvars in Deferred](https://twistedmatrix.com/trac/ticket/9807)
- [Support contextvars in DelayedCall](https://twistedmatrix.com/trac/ticket/9824)
- [Contextvars support does not support `.reset`](https://twistedmatrix.com/trac/ticket/10301)